### PR TITLE
[Merged by Bors] - feat: iterated derivative of a continuous multilinear map

### PR DESCRIPTION
**Pull request by sgouezel**
Consider a multilinear map `f : E^n -> F`. Then its `k`-th derivative can be written as follows, given a basepoint `x : E^n` and `k` vectors `v_1, ..., v_k : E^n`:
```
D^k f (x; v_1, ..., v_k) = \sum f (x_1, (v_{e.symm 2})_2, x_3, ..., (v_{e.symm n))_n)
```
where in the sum at each index `i` one uses either the `i`-th coordinate of `x`, or the `i`-th coordinate of one of the vectors `v_j`, and each `v_j` has to appear exactly once in this expression. The sum is parameterized by the embeddings `e` of `Fin k` in `Fin n`.

This expression is a little bit cumbersome to write down and to use, but it gives nice explicit bounds on `||D^k f||` that are useful to study Fourier transforms. 

The PR introduces the formal definition of each term in the sum (as an object which is multilinear in the coordinates of `x` outside the range of `e`, and in the `v_j`), and then proves that the sum is indeed the iterated derivative of `f`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by loefflerd — COMMENTED** [review]
I've left some comments, most of which are golf / pedantry – the overall design looks spot-on.

File: Mathlib/Analysis/Calculus/FDeriv/Analytic.lean [hunk 1]
```diff
@@ -347,6 +350,88 @@ protected theorem hasFDerivAt [DecidableEq ι] : HasFDerivAt f (f.linearDeriv x)
...
+      simp only [hke, hkf, ↓reduceDite]
+      simp only [Pi.compRightL, ContinuousLinearMap.coe_mk', LinearMap.coe_mk, AddHom.coe_mk]
+      rw [Function.update_noteq]
+      contrapose! hke
+      have hk : k = (Equiv.embeddingFinSucc n ι e).2 := Subtype.ext_iff_val.1 hke
+      rw [hk, Equiv.embeddingFinSucc_snd e]
+      apply Set.mem_range_self
+  · intro n _hn
+    apply continuous_finset_sum _ (fun e _he ↦ ?_)
+    exact (ContinuousMultilinearMap.coe_continuous _).comp (ContinuousLinearMap.continuous _)
```
[the remark this answers is not in the document]
Comment (sgouezel) [reply Mathlib/Analysis/Calculus/FDeriv/Analytic.lean:420]: Lokks a little bit less readable to me, so I'll keep the original version.

File: Mathlib/Analysis/Calculus/FDeriv/Analytic.lean [hunk 2]
```diff
@@ -347,6 +350,88 @@ protected theorem hasFDerivAt [DecidableEq ι] : HasFDerivAt f (f.linearDeriv x)
...
+    suffices H : curryLeft (f.iteratedFDeriv (Nat.succ n) x) = (∑ e : Fin n ↪ ι,
+          ((iteratedFDerivComponent f e.toEquivRange).linearDeriv
+            (Pi.compRightL 𝕜 _ Subtype.val x)) ∘L (Pi.compRightL 𝕜 _ Subtype.val)) by
+      have A : HasFDerivAt (f.iteratedFDeriv n) (∑ e : Fin n ↪ ι,
+          ((iteratedFDerivComponent f e.toEquivRange).linearDeriv (Pi.compRightL 𝕜 _ Subtype.val x))
+            ∘L (Pi.compRightL 𝕜 _ Subtype.val)) x := by
+        apply HasFDerivAt.sum (fun s _hs ↦ ?_)
+        exact (ContinuousMultilinearMap.hasFDerivAt _ _).comp x (ContinuousLinearMap.hasFDerivAt _)
+      rwa [← H] at A
+    ext v m
```
[the remark this answers is not in the document]
Comment (sgouezel) [reply Mathlib/Analysis/Calculus/FDeriv/Analytic.lean:386]: Agreed.

File: Mathlib/LinearAlgebra/Multilinear/Basic.lean [hunk 3]
```diff
@@ -1004,6 +1022,70 @@ variable [CommSemiring R] [∀ i, AddCommMonoid (M₁ i)] [∀ i, AddCommMonoid
...
+    simp
+  · rw [update_noteq, update_noteq]
+    · rintro rfl
+      simp at hne
+    · exact hne.symm
+
+/-- One of the components of the iterated derivative of a multilinear map. Given a bijection `e`
+between a type `α` (typically `Fin k`) and a subset `s` of `ι`, this component is a multilinear map
+of `k` vectors `v₁, ..., vₖ`, mapping them
+to `f (x₁, (v_{e.symm 2})₂, x₃, ...)`, where at indices `i` in `s` one uses the `i`-th coordinate of
```
[the remark this answers is not in the document]
Comment (sgouezel) [reply Mathlib/LinearAlgebra/Multilinear/Basic.lean:1050]: I didn't get this one. What do you suggest?

**Comment by sgouezel** [comment]
Thanks a lot for the review!

File: Mathlib/LinearAlgebra/Multilinear/Basic.lean (hunk 3 above)
[the remark this answers is not in the document]
Comment (loefflerd) [reply Mathlib/LinearAlgebra/Multilinear/Basic.lean:1050]: Just that the line break after `them` is oddly positioned.

[the remark this answers is not in the document]
Comment (sgouezel) [reply Mathlib/LinearAlgebra/Multilinear/Basic.lean]: Fixed, thanks!

**Comment by loefflerd** [comment]
Looks great now! 

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Comment by loefflerd** [comment]
Looks like it failed because `gcongr` got better – definitely the best kind of build failure :smile:

**Pull request closed without merging**